src/engine/strat_engine/keys.rs

```diff
@@ -0,0 +1,421 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::{
+    convert::TryFrom,
+    ffi::CString,
+    fs::File,
+    io::{self, Read},
+    os::unix::io::{FromRawFd, RawFd},
+    str,
+};
+
+use libc::{syscall, SYS_add_key, SYS_keyctl};
+
+use devicemapper::Bytes;
+use libcryptsetup_rs::SafeMemHandle;
+
+use crate::{
+    engine::{
+        engine::{KeyActions, MAX_STRATIS_PASS_SIZE},
+        strat_engine::names::KeyDescription,
+        types::{DeleteAction, KeySerial, MappingCreateAction, SizedKeyMemory},
+    },
+    stratis::{ErrorEnum, StratisError, StratisResult},
+};
+
+/// Get the ID of the persistent root user keyring and attach it to
+/// the session keyring.
+fn get_persistent_keyring() -> StratisResult<KeySerial> {
+    // Attach persistent keyring to session keyring
+    match unsafe {
+        syscall(
+            SYS_keyctl,
+            libc::KEYCTL_GET_PERSISTENT,
+            0,
+            libc::KEY_SPEC_SESSION_KEYRING,
+        )
+    } {
+        i if i < 0 => Err(io::Error::last_os_error().into()),
+        i => Ok(i as KeySerial),
+    }
+}
+
+/// Search for the given key description in the persistent root keyring.
+/// Returns the key ID or nothing if it was not found in the keyring.
+fn search_key(
+    keyring_id: KeySerial,
+    key_desc: &KeyDescription,
+) -> StratisResult<Option<KeySerial>> {
+    let key_desc_cstring = CString::new(key_desc.to_system_string()).map_err(|_| {
+        StratisError::Engine(
+            ErrorEnum::Invalid,
+            "Invalid key description provided".to_string(),
+        )
+    })?;
+
+    let key_id = unsafe {
+        syscall(
+            SYS_keyctl,
+            libc::KEYCTL_SEARCH,
+            keyring_id,
+            concat!("user", "\0").as_ptr(),
+            key_desc_cstring.as_ptr(),
+            0,
+        )
+    };
+    if key_id < 0 {
+        if unsafe { *libc::__errno_location() } == libc::ENOKEY {
+            Ok(None)
+        } else {
+            Err(io::Error::last_os_error().into())
+        }
+    } else {
+        Ok(Some(key_id as KeySerial))
+    }
+}
+
+/// Read a key with the provided key description into safely handled memory if it
+/// exists in the keyring.
+///
+/// The return type with be a tuple of an `Option` and a keyring id. The `Option`
```

Comment:
Consider "The return type _will_ be a tuple..."?

Reply:
@mulkieran Is this change okay to squash?

Reply:
Yes, squash that change.